providers/common/ciphers/aes_gcm_hw.c

```diff
@@ -0,0 +1,278 @@
+                                 const unsigned char *iv, size_t ivlen);
+static int generic_aes_gcm_aad_update(PROV_AES_GCM_CTX *ctx,
+                                      const unsigned char *aad, size_t aad_len);
+static int generic_aes_gcm_cipher_update(PROV_AES_GCM_CTX *ctx,
+                                         const unsigned char *in, size_t len,
+                                         unsigned char *out);
+static int generic_aes_gcm_cipher_final(PROV_AES_GCM_CTX *ctx,
+                                        unsigned char *tag);
+static int generic_aes_gcm_one_shot(PROV_AES_GCM_CTX *ctx,
+                                    unsigned char *aad, size_t aad_len,
+                                    const unsigned char *in, size_t in_len,
+                                    unsigned char *out,
+                                    unsigned char *tag, size_t tag_len);
+
+#define SET_KEY_CTR_FN(fn_set_enc_key, fn_aes, fn_ctr)                         \
+    fn_set_enc_key(key, keylen * 8, &ctx->ks.ks);                              \
+    CRYPTO_gcm128_init(&ctx->gcm, &ctx->ks, (block128_f)fn_aes);               \
+    ctx->ctr = (ctr128_f)fn_ctr;                                               \
+    ctx->key_set = 1;
+
+#if defined(AESNI_CAPABLE)
+
+/* AES-NI section */
+
+static int aesni_gcm_init_key(PROV_AES_GCM_CTX *ctx,
+                              const unsigned char *key, size_t keylen)
+{
+    SET_KEY_CTR_FN(aesni_set_encrypt_key, aesni_encrypt,
+                   aesni_ctr32_encrypt_blocks);
+    return 1;
+}
+
+static const PROV_AES_GCM_HW aesni_gcm = {
+    aesni_gcm_init_key,
+    generic_aes_gcm_setiv,
+    generic_aes_gcm_aad_update,
+    generic_aes_gcm_cipher_update,
+    generic_aes_gcm_cipher_final,
+    generic_aes_gcm_one_shot
+};
+
+const PROV_AES_GCM_HW *PROV_AES_HW_gcm(size_t keybits)
+{
+    return AESNI_CAPABLE ? &aesni_gcm : &aes_gcm;
+}
+
+#elif defined(AES_ASM) && (defined(__sparc) || defined(__sparc__))
+
+/* Fujitsu SPARC64 X support */
+
+static int t4_aes_gcm_init_key(PROV_AES_GCM_CTX *ctx,
+                               const unsigned char *key, size_t keylen)
+{
+    ctr128_f ctr;
+
+    switch (keylen) {
+    case 16:
+        ctr = (ctr128_f)aes128_t4_ctr32_encrypt;
+        break;
+    case 24:
+        ctr = (ctr128_f)aes192_t4_ctr32_encrypt;
+        break;
+    case 32:
+        ctr = (ctr128_f)aes256_t4_ctr32_encrypt;
+        break;
+    default:
+        return 0;
+    }
+
+    SET_KEY_CTR_FN(aes_t4_set_encrypt_key, aes_t4_encrypt, ctr);
+    return 1;
+}
+
+static const PROV_AES_GCM_HW t4_aes_gcm = {
+    t4_aes_gcm_init_key,
+    generic_aes_gcm_setiv,
+    generic_aes_gcm_aad_update,
+    generic_aes_gcm_cipher_update,
+    generic_aes_gcm_cipher_final,
+    generic_aes_gcm_one_shot
+};
+const PROV_AES_GCM_HW *PROV_AES_HW_gcm(size_t keybits)
+{
+    return SPARC_AES_CAPABLE ? &t4_aes_gcm : &aes_gcm;
+}
+
+#elif defined(OPENSSL_CPUID_OBJ) && defined(__s390__)
+# include "aes_gcm_s390x.inc"
```

Comment:
Would it be better to keep this as a ".c" file so that editors detect it as C?

Reply:
ok - was trying to make it clear that it was embedded. Added a comment at the top instead.